### PR TITLE
add: output improvements

### PR DESCRIPTION
**Pull request by casperdcl**
`dvc add` output improvements

- [x] slight wording updates
- [x] add progress for `stage.create()` (cleared upon completion)
- [x] leave some stats behind?
- [x] fix failing tests
- [x] fix md3sum progress not clearing properly
- [x] progress-aware `logger.info`
  + [x] fix/update tests
- [x] `To track the changes with git, run`: prints once rather than per-file
- [x] `Computing md5 for a large number of files. This is only done once.` message replaced with `Computing md5 for a large number of files` inside cleared away progress bars
- [x] `Saving information to '*.dvc'.` messages removed (since files are already mentioned in final commit reminder)
  + [ ] make sure all other calls to `stage.dump()` do not need this verbosity

old:
[asciicast](https://asciinema.org/a/m9PX4jUPGZaE1gWp5eWbgEns3?speed=2)
new:
[asciicast](https://asciinema.org/a/VC7HfMbG6XxkesVy5sRduHLNN?speed=4)

## related
- #2545
- https://github.com/iterative/dvc/pull/2546#discussion_r337037137

## for future issue/PR
- #2658

### CODE REVIEW

**Comment by efiop** [comment]
@casperdcl Please use your forked repo to work on such PRs. We have some stuff to improve until we allow in-upstream branches, since travis runs two checks and one always fails.  Plus it also helps keep the upstream tidy :)

File: dvc/command/add.py [hunk 1]
```diff
@@ -14,38 +16,44 @@ class CmdAdd(CmdBase):
...
-                    recursive=self.args.recursive,
-                    no_commit=self.args.no_commit,
-                    fname=self.args.file,
+                raise RecursiveAddingWhileUsingFilename(
+                    "can't use '--file' with multiple targets"
                 )
 
-        except DvcException:
-            logger.exception("failed to add file")
+            with Tqdm(
```
Comment (efiop) [inline dvc/command/add.py]: It would be really nice to see before and after comparison, something like you've done with asciinema before. It really helps visualize this stuff.

**Comment by casperdcl** [comment]
@efiop here's a recording of before and after. It hihglights omse remaining issues.

[asciicast](https://asciinema.org/a/m9PX4jUPGZaE1gWp5eWbgEns3?speed=2)

1. there was a problem with md3sum progress not clearing properly (happens e.g. at 0:50) which I've now fixed
2. `logger.info` is not progress-aware (`logger.debug`, `logger.error` are aware). This needs fixing (happens e.g. at 2:40)
3. `To track the changes with git, run:` really should print at the end with a list of dvc files rather than one at a time with lots of empty newlines.
4. Still need to add some stats

**Comment by casperdcl** [comment]
@efiop new version:

[asciicast](https://asciinema.org/a/bOzNFCfHJ7ihuQTCeTN0T4QbC)

**Comment by casperdcl** [comment]
main problem is `dvc.repo.scm_context.scm_context` has a `repo.scm.remind_to_track()` which should ideally be collected and delayed until after all `add`s get executed.

**Comment by efiop** [comment]
@casperdcl You can simply make `dvc/repo/add.py` accept multiple targets and then process them accordingly, instead of doing that on CLI level in `dvc/command/add.py`. That way even `scm_context` will work properly automatically.

**Comment by casperdcl** [comment]
new version with only final commit reminder message:

[asciicast](https://asciinema.org/a/DCroBSXaXjceBdWXc38Z7FGMO?speed=2)

**Comment by casperdcl** [comment]
@efiop I think

1. `Computing md5 for a large number of files. This is only done once.` should be removed and instead the (cleared away) md5sum bars should have a message `Computing md5 for a large number of files`
2. The `Saving information to '*.dvc'.` messages should be removed

**Comment by shcheklein** [comment]
@casperdcl @efiop if remove stuff (and bars after add is done) we need to show some message at least - one per file? Something like `Adding foo1 ... ✔️`. Or one message `Added 10 files in 1234 seconds`? It looks weird when we don't have anything at all at the (like for checkout/pull now).

**Comment by casperdcl** [comment]
new version with decreased verbosity:

[asciicast](https://asciinema.org/a/00aV99LP3JtPE8jRULDI80Vsu)

Only thing missing is final stats

**Comment by casperdcl** [comment]
with some stats:

[asciicast](https://asciinema.org/a/VC7HfMbG6XxkesVy5sRduHLNN?speed=2)

**Comment by casperdcl** [comment]
@efiop think this is mergable as-is and then open a new issue/PR pair for refinement.

File: dvc/repo/add.py [hunk 2]
```diff
@@ -3,52 +3,75 @@
...
-
 logger = logging.getLogger(__name__)
 
 
 @locked
 @scm_context
-def add(repo, target, recursive=False, no_commit=False, fname=None):
+def add(
+    repo,
+    target_list,
```
Comment (efiop) [inline dvc/repo/add.py]: let's just call it
```suggestion
    targets,
```
same as we have in other places.

File: dvc/command/add.py [hunk 3]
```diff
@@ -54,7 +57,7 @@ def add_parser(subparsers, parent_parser):
         help="Don't put files/directories into cache.",
     )
     add_parser.add_argument(
-        "-f", "--file", help="Specify name of the DVC-file it generates."
+        "-f", "--file", help="Specify name of the generated DVC file."
```
Comment (efiop) [inline dvc/command/add.py]:
```suggestion
        "-f", "--file", help="Specify name of the generated DVC-file."
```

File: dvc/command/add.py [hunk 4]
```diff
@@ -14,38 +15,40 @@ class CmdAdd(CmdBase):
...
+            self.repo.add(
+                self.args.targets,
+                recursive=self.args.recursive,
+                no_commit=self.args.no_commit,
+                fname=self.args.file,
+                progress=True,
+            )
+
+        except DvcException as err:
+            logger.exception("{}:{}".format(type(err).__name__, err))
```
Comment (efiop) [inline dvc/command/add.py]: Why do you have to `format` exception here?

File: dvc/command/add.py [hunk 5]
```diff
@@ -14,38 +15,40 @@ class CmdAdd(CmdBase):
...
-                raise DvcException("can't use '--file' with multiple targets")
-
-            for target in self.args.targets:
-                self.repo.add(
-                    target,
-                    recursive=self.args.recursive,
-                    no_commit=self.args.no_commit,
-                    fname=self.args.file,
+                raise RecursiveAddingWhileUsingFilename(
+                    "can't use '--file' with multiple targets"
```
Comment (efiop) [inline dvc/command/add.py]: Isn't it the same as default message in `RecursiveAddingWhileUsingFilename`?

File: dvc/logger.py [hunk 6]
```diff
@@ -170,15 +175,25 @@ def setup(level=logging.INFO):
...
+                "exclude_errors": {"()": ExcludeErrorsFilter},
+                "exclude_info": {"()": ExcludeInfoFilter},
+            },
             "formatters": {"color": {"()": ColorFormatter}},
             "handlers": {
-                "console": {
+                "console_info": {
+                    "class": "dvc.logger.LoggerHandler",
+                    "level": "INFO",
+                    # "formatter": "color",
```
Comment (efiop) [inline dvc/logger.py]: ?

File: dvc/logger.py [hunk 7]
```diff
@@ -170,15 +175,25 @@ def setup(level=logging.INFO):
     logging.config.dictConfig(
         {
             "version": 1,
-            "filters": {"exclude_errors": {"()": ExcludeErrorsFilter}},
+            "filters": {
```
Comment (efiop) [inline dvc/logger.py]: Could you explain what is going on here in general? The commit message doesn't have any explanation.

File: dvc/repo/add.py [hunk 8]
```diff
@@ -3,52 +3,75 @@
...
-                purple=colorama.Fore.MAGENTA,
-                cyan=colorama.Fore.CYAN,
-                nc=colorama.Style.RESET_ALL,
-                target=target,
-            )
-        )
+    stages_list = []
+    with Tqdm(
+        total=len(target_list),
+        desc="Adding",
```
Comment (efiop) [inline dvc/repo/add.py]: It is weird that this is only shown for 0%, after which it simply shows file names.

File: dvc/repo/add.py [hunk 9]
```diff
@@ -3,52 +3,75 @@
...
 
-        stage.dump()
+                if not no_commit:
+                    stage.commit()
 
-    return stages
+                stage.dump()
+
+            stages_list += stages
+        pbar.bar_format = pbar.BAR_FMT_DEFAULT.replace("|{bar:10}|", " ")
```
Comment (efiop) [inline dvc/repo/add.py]: Could you explain what this is for please?

**Comment by efiop** [comment]
> The recording clearly shows that md5sums only take about 10% of the total time. There must be other more time consuming functions.

@casperdcl Could you cProfile it or just pdb it to see what is taking so long?

File: dvc/repo/add.py (hunk 2 above)
Comment (casperdcl) [reply dvc/repo/add.py]: I have to avoid reusing the name `targets` as it's redefined later:

https://github.com/iterative/dvc/blob/8f9369065350a0d6f70858b3ff9e864470af34d9/dvc/repo/add.py#L43

at least one of these vars has to have a different name

File: dvc/command/add.py (hunk 4 above)
Comment (casperdcl) [reply dvc/command/add.py]: no particular reason. would you prefer string addition or `':'.join` instead?

File: dvc/command/add.py (hunk 5 above)
Comment (casperdcl) [reply dvc/command/add.py]: essentially sort of yes. would you prefer

1. `can't use '--file' with multiple targets`
2. `using fname with recursive is not allowed.`
3. something else

File: dvc/logger.py (hunk 6 above)
Comment (casperdcl) [reply dvc/logger.py]: see below explanation

File: dvc/logger.py (hunk 7 above)
Comment (casperdcl) [reply dvc/logger.py]: this was needed as previously only `logging.DEBUG` and `logging.WARNING` were handled by `dvc.logger.LoggerHandler`. However this change ensures that `logging.INFO` is also handled by our class. The custom class is required to ensure that `logger.info()` interleaves properly with `dvc.progress.Tqdm()`.

I should've described this commit as "progress-aware `logger.info`"

File: dvc/repo/add.py (hunk 8 above)
Comment (casperdcl) [reply dvc/repo/add.py]: yes but after the bar is completed it goes back to showing the initial message (i.e. "Adding").
I was thinking of changing it to just "Add" from the start though.

File: dvc/repo/add.py (hunk 9 above)
Comment (casperdcl) [reply dvc/repo/add.py]: just removing the progress bar bit upon completion as it seems unnecessary. would expect we need a separate PR later to actually fully sort out exactly what stats we want displayed upon completion so this line will change in future.

**Comment by casperdcl** [comment]
> Could you cProfile it or just pdb it to see what is taking so long?

yes but would prefer to do so in a separate PR. maybe open an issue after this is merged?

**Comment by casperdcl** [comment]
@efiop the main problem with this PR (maybe it's not a problem - I haven't had time to check and it might be much quicker for you to check) is

> make sure all other calls to `stage.dump()` do not need this verbosity

I made `stage.dump()` less verbose but that may not be wanted in all the the 13 places it's used. It's likely at least some of those uses will at least need the same progress wrapping that is in this PR.

File: dvc/repo/add.py (hunk 2 above)
Comment (efiop) [reply dvc/repo/add.py]: @casperdcl let's rename that one, so the API is nice and uniform names.

File: dvc/command/add.py (hunk 4 above)
Comment (efiop) [reply dvc/command/add.py]: @casperdcl Not sure I follow. `logger.exception` will automatically print last exception and format it with `:`, why do we need to format this separately?

File: dvc/command/add.py (hunk 5 above)
Comment (efiop) [reply dvc/command/add.py]: @casperdcl Would it hurt if we combine them into one message? Smth like
```
using fname with multiple targets or `-R|--recursive` is not allowed
```

File: dvc/logger.py (hunk 6 above)
Comment (efiop) [reply dvc/logger.py]: Yeah, but why do we have a commented-out line here?

File: dvc/command/add.py (hunk 4 above)
Comment (casperdcl) [reply dvc/command/add.py]: ah right. so just `logger.exception(err)`?

File: dvc/logger.py (hunk 6 above)
Comment (casperdcl) [reply dvc/logger.py]: ah yes, it can be removed as it's unneeded. on the other hand the colour formatter seems to pass through `INFO` without altering so might just uncomment for future-proofing.

File: dvc/repo/add.py (hunk 9 above)
Comment (jorgeorpinel) [reply dvc/repo/add.py]: Hi. On this, and maybe unrelated to this PR but, are we sure we want the progress bars to go away after they're done, in general? It's kind of nice to be able to see what happened in the past when the command is done, for example if you run a long `dvc add` and go away from your computer for a while. Just a thought.

**Comment by jorgeorpinel** [comment]
> ... would prefer to do so in a separate PR.

My 2c here: It's probably best to profile before merging this, if it seems like the change it's slowing down the tool.

File: dvc/repo/add.py (hunk 9 above)
Comment (casperdcl) [reply dvc/repo/add.py]: yes in general I think it's not worth removing progress bars as aggressively as we've currently done but in this instance it's only remving the bar bit and not the stats 

```
100%|███████████|Adding 10000/10000 [00:43<00:00, 229.00file/s]
```

becomes

```
100% Adding             10000/10000 [00:43<00:00, 229.00file/s]
```

File: dvc/command/add.py [hunk 10]
```diff
@@ -14,38 +15,38 @@ class CmdAdd(CmdBase):
...
+
+        except DvcException as err:
+            logger.exception("{}:{}".format(type(err).__name__, err))
             return 1
         return 0
 
 
 def add_parser(subparsers, parent_parser):
-    ADD_HELP = "Take data files or directories under DVC control."
+    add_help = "Add data files or directories to DVC control."
```
[the remark this answers is not in the document]
Comment (casperdcl) [reply dvc/command/add.py]: I was fixing DeepSource (caps for non-global-const) - if there are other violations technically should fix those too...

File: dvc/stage.py [hunk 11]
```diff
@@ -663,7 +663,7 @@ def dump(self):
 
         self._check_dvc_filename(fname)
 
-        logger.info(
+        logger.debug(
             "Saving information to '{file}'.".format(file=relpath(fname))
```
Comment (jorgeorpinel) [inline dvc/stage.py]: And one last question: Is this really a debug output? I also find that message useful in the output.

File: dvc/command/add.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (casperdcl) [reply dvc/command/add.py]: hmm "Take directories under control" isn't good English though. Would it be a pain to update the docs? More accurately I'd prefer "Select files for DVC to track" or similar. Note that `git add --help` gives "Add file contents to the index."

[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply dvc/command/add.py]: Ah, interesting. Well, I definitely see lots of these `{CMD}_HELP` non-global constants everywhere. Let's see what @efiop decides.

File: dvc/command/add.py [hunk 12]
```diff
@@ -54,7 +55,7 @@ def add_parser(subparsers, parent_parser):
         help="Don't put files/directories into cache.",
     )
     add_parser.add_argument(
-        "-f", "--file", help="Specify name of the DVC-file it generates."
+        "-f", "--file", help="Specify name of the generated DVC-file."
```
[the remark this answers is not in the document]
Comment (casperdcl) [reply dvc/command/add.py]: "Specify name of the output DVC-file?"

File: dvc/exceptions.py [hunk 13]
```diff
@@ -226,7 +226,7 @@ def __init__(self, path, cause=None):
 class RecursiveAddingWhileUsingFilename(DvcException):
     def __init__(self):
         super(RecursiveAddingWhileUsingFilename, self).__init__(
-            "using fname with recursive is not allowed."
+            "cannot use `fname` with multiple targets or `-R|--recursive`"
```
[the remark this answers is not in the document]
Comment (casperdcl) [reply dvc/exceptions.py]: unlikely; think it'll wind up following a colon after the exception type.

File: dvc/stage.py (hunk 11 above)
Comment (casperdcl) [reply dvc/stage.py]: the thing is the tracked file name will be printed in the progress bar itself, while the dvc file names will all be printed at the end. This means this message should no longer be required. I left it in for 1. debugging and 2. not sure if the 12 other uses of `stage.dump()` actually need it to be `logger.info`. This is the only real issue I have with this PR atm.

File: dvc/repo/add.py (hunk 9 above)
Comment (casperdcl) [reply dvc/repo/add.py]: still could do with more tidying in a different PR I think: `<00:00` is useless; total size in bytes would be nice; "Adding" should be "Added;" etc...

File: dvc/command/add.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply dvc/command/add.py]: Hmm I guess you're right. Maybe just replace "take" for "place". But if we're changing this we might as well re-think the entire wording. How about "Track data files or directories with DVC." ? Any thoughts @shcheklein?

And it's no problem to update docs, I just opened dvc.org/issues/719 for that effect, feel fee to self-assign if you'd like to also address it but someone else will get it if not.

p.s. I think we don't want to use the word "add" to define `dvc add`, even if Git doesn't care. 😋

File: dvc/command/add.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply dvc/command/add.py]: "output" has a specific meaning in the context of DVC, would rather avoid using here. I don't think it's a problem to have a long _ish_ help message like the one I proposed. 🙂

File: dvc/exceptions.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply dvc/exceptions.py]: I would still capitalize after `:` but OK, up to you. I'm resolving this one. 🙂

File: dvc/stage.py (hunk 11 above)
Comment (jorgeorpinel) [reply dvc/stage.py]: Cool. Feel fee to ask @efiop and/or resolve this conversation when you see fit. Thanks.

File: dvc/repo/add.py (hunk 9 above)
Comment (jorgeorpinel) [reply dvc/repo/add.py]: Good ideas. Feel fee to open a separate issue with these concerns (or find an existing ones to add your comments) if this is not the main purpose of the present PR.

File: dvc/command/add.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply dvc/command/add.py]: Hi. I see no change here. Please use my latest suggestion, "Track data files or directories with DVC." Thanks

[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply dvc/command/add.py]: (This is the only remaining part of my review, all the other comments are just questions.)

File: dvc/command/add.py (hunk 4 above)
Comment (efiop) [reply dvc/command/add.py]: @casperdcl if you only need to print the exception, then `logger.exception()`, err will be parsed automatically.

File: dvc/repo/add.py (hunk 9 above)
Comment (efiop) [reply dvc/repo/add.py]: @casperdcl Honestly, this is a relly obscure piece of code. At the minimum we need to put a comment there.

File: dvc/command/add.py [hunk 14]
```diff
@@ -14,49 +15,51 @@ class CmdAdd(CmdBase):
...
         except DvcException:
-            logger.exception("failed to add file")
+            logger.exception("")
             return 1
         return 0
 
 
 def add_parser(subparsers, parent_parser):
-    ADD_HELP = "Take data files or directories under DVC control."
+    add_help = "Track data files or directories with DVC."
```
Comment (efiop) [inline dvc/command/add.py]: This var name renaming stuff here and down below doesn't really belong to this pr :slightly_smiling_face: Also, as @jorgeorpinel noticed, if we modify these, we should also modify all the others in dvc/command/*.py, to keep the code uniformed.

File: dvc/command/add.py [hunk 15]
```diff
@@ -14,49 +15,51 @@ class CmdAdd(CmdBase):
...
-                    recursive=self.args.recursive,
-                    no_commit=self.args.no_commit,
-                    fname=self.args.file,
-                )
+            self.repo.add(
+                self.args.targets,
+                recursive=self.args.recursive,
+                no_commit=self.args.no_commit,
+                fname=self.args.file,
+                progress=True,
```
Comment (efiop) [inline dvc/command/add.py]: I don't think we need this one. Only disabling the progress bar won't help anyone much, as he will be bombarded by our other messages anyway. We will need to come up with a silencing strategy for dvc in general later. (could do that through logger right now, though).
```suggestion
```

File: dvc/command/base.py [hunk 16]
```diff
@@ -57,7 +57,7 @@ def default_targets(self):
 
     # Abstract methods that have to be implemented by any inheritance class
     def run(self):
-        pass
+        raise NotImplementedError("Abstract")
```
Comment (efiop) [inline dvc/command/base.py]:
```suggestion
        raise NotImplementedError
```
should be enough and conforms to the rest of the codebase :slightly_smiling_face: E.g. `dvc/remote/base.py`

File: dvc/command/add.py (hunk 14 above)
Comment (casperdcl) [reply dvc/command/add.py]: ok I'll revert the `ADD_HELP` for consistency. I'll leave the `add_parser` => `parser` as that fixes a warning and doesn't really introduce inconsistency.

File: dvc/command/add.py (hunk 15 above)
Comment (efiop) [reply dvc/command/add.py]: @casperdcl `progress` is still here, why is this one marked as resolved?

File: dvc/repo/add.py [hunk 17]
```diff
@@ -3,52 +3,72 @@
...
-        stage.dump()
+                if not no_commit:
+                    stage.commit()
 
-    return stages
+                stage.dump()
+
+            stages_list += stages
+        # remove filled bar bit of progress, leaving stats.
+        # TODO: tidy in next PR.
```
Comment (efiop) [inline dvc/repo/add.py]: Please don't leave `TODO`s in the code. If issue is significant, just create a github issue and assign to yourself.

File: dvc/command/add.py (hunk 15 above)
Comment (casperdcl) [reply dvc/command/add.py]: resolved locally, haven't pushed yet because it's related to a comment (below). Not sure if it went through the github interface, though. trying to recomment...

File: dvc/repo/add.py (hunk 17 above)
Comment (casperdcl) [reply dvc/repo/add.py]: fine opened #2658

**Comment by efiop** [comment]
@casperdcl 

```
#!/bin/bash                                     
                                                
set -e                                          
set -x                                          
                                                
rm -rf myrepo                                   
mkdir myrepo                                    
cd myrepo                                       
dvc init --no-scm                               
                                                
for i in $(seq 1 102); do                      
    dd if=/dev/urandom of=data$i bs=1M count=100
done                                            
                                                
dvc add data*                                   
```

not seeing any progress bars at all.

**Comment by casperdcl** [comment]
@efiop that's because I rebased onto master before the silly `isatty` fix, just rebased again; should work now

File: dvc/command/add.py (hunk 15 above)
Comment (efiop) [reply dvc/command/add.py]: Ah, got it. Sorry :)

Comment (casperdcl) [reply dvc/command/add.py]: yes I think I hit Ctrl+Enter which started a (pending) review rather than adding a single comment so my message didn't go through.

File: dvc/repo/add.py [hunk 18]
```diff
@@ -3,52 +3,70 @@
...
 
-
 logger = logging.getLogger(__name__)
 
 
 @locked
 @scm_context
-def add(repo, target, recursive=False, no_commit=False, fname=None):
+def add(
+    repo, targets, recursive=False, no_commit=False, fname=None, progress=False
```
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/repo/add.py]: @casperdcl I would remove it totally. I'm not seeing any need for it.

File: dvc/command/add.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (casperdcl) [reply dvc/command/add.py]: reverted (possibly to fix in another PR)

**Comment by casperdcl** [comment]
@efiop not sure why you've started squash-merging PRs. Think history (esp for complex PRs like this one) is really important for tracking future issues...

**Comment by efiop** [comment]
@casperdcl This type of PR history would've been squashed anyway, because it is a total mess with "fix this" and "fix that" commits, which are a total nightmare to browse the history with. I usually prefer rebasing and squashing commits to make them nice and granular, but many contributors are used to just piling commits on top of each other, so I've decided to just squash everything to make it look nice in the history, while granular commits are available in the original PR on github along with all the context from the comments.

**Comment by casperdcl** [comment]
imho nothing beats `git checkout feature && git rebase master && git checkout master && git merge --no-ff feature-branch`

**Comment by jorgeorpinel** [comment]
@casperdcl for commit message consistency please refer to https://dvc.org/doc/user-guide/contributing/core#commit-message-format-guidelines. Thanks

**Comment by casperdcl** [comment]
@jorgeorpinel sure; fine with most of it except don't agree with "Use `dvc` in a general case"

**Comment by jorgeorpinel** [comment]
Feel fee to submit a PR with a suggested update to that page and I'll tag the core dev team to discuss. 🙂
